spectral: v2019-03-03 -> v2019-08-15

colemickens opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Update `spectral`.

Works for me!

cc: @dtzWill

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

colemickens committed:
spectral: 2019-03-03 -> unstable-2019-08-15

rnhmjoj commented:
Can you remove the wrapper and use `wrapQtAppsHook` instead? See for example https://github.com/NixOS/nixpkgs/pull/54525/commits/51d78034a1db78f8ce0ea3ba6fa20be590ed0ca2

colemickens commented:
Done. I think. It works for me anyway.

rnhmjoj reviewed (changes requested):
I think this whole `let` block can be removed.

colemickens commented:
Yes, of course, thanks. Pushed again.

rnhmjoj commented:
The wrapper is working but I still can't run the application without installing it:
```
[nix-shell:~/.cache/nix-review/pr-65354-4]$ results/matrique/bin/spectral
DPI: 1
Cannot mix incompatible Qt library (version 0x50c00) with this library (version 0x50c03)
Aborted
```
@ttuegel do you know what's happening here?

ttuegel commented on pkgs/applications/networking/instant-messengers/spectral/default.nix:
```diff
@@ -8,35 +8,20 @@
-  ];
-
-in stdenv.mkDerivation rec {
+stdenv.mkDerivation rec {
```
The preferred way to do this is use `mkDerivation` instead of `stdenv.mkDerivation`, so that you do not need to call `wrapQtAppsHook` directly. `mkDerivation` also sets some Qt-specific build flags. Please refer to [the manual](https://hydra.nixos.org/build/96804884/download/1/nixpkgs/manual.html#sec-language-qt) for more information.

colemickens replied:
Hm, I checked the docs, it seems really confusing to name them the same thing. Plus, I assume `mkDerivation` isn't soley for Qt apps, or is it?

I can take a look at this later and try to get it working with the more succinct syntax. Thanks for pointing this out and linking the docs!

ttuegel commented:
> do you know what's happening here?

There is likely something incompatible installed in your user or system environment. What is `QT_PLUGIN_PATH` set to in your shell, and what are the contents of all the directories listed there?

rnhmjoj commented:
@ttugel That's that I thought but the error persist even after unsetting QT_PLUGIN_PATH and QML_IMPORT_PATH.

ttuegel commented:
@rnhmjoj Could you try to run it under `strace` and find out what file or library it's trying to open? There are several other ways it can attempt to load things from the environment. :slightly_frowning_face:

rnhmjoj commented:
So, I can see it's loading both qtbase 5.12.0 and 5.12.3. I think it found about it by looking inside my nix profile:
```
lstat("/nix/var/nix/profiles", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/nix/var/nix/profiles/per-user", {st_mode=S_IFDIR|S_ISVTX|0777, st_size=4096, ...}) = 0
lstat("/nix/var/nix/profiles/per-user/rnhmjoj", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/nix/var/nix/profiles/per-user/rnhmjoj/profile", {st_mode=S_IFLNK|0777, st_size=16, ...}) = 0
lstat("/nix/var/nix/profiles/per-user/rnhmjoj/profile-191-link", {st_mode=S_IFLNK|0777, st_size=60, ...}) = 0
lstat("/nix", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/nix/store", {st_mode=S_IFDIR|S_ISVTX|0775, st_size=5107712, ...}) = 0
lstat("/nix/store/2hk43911y2addhkcai4zblnihmwrvsj6-user-environment", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/nix/store/2hk43911y2addhkcai4zblnihmwrvsj6-user-environment/lib", {st_mode=S_IFLNK|0777, st_size=65, ...}) = 0
lstat("/nix", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/nix/store", {st_mode=S_IFDIR|S_ISVTX|0775, st_size=5107712, ...}) = 0
lstat("/nix/store/7bvh4ydpza5cfhd7rw7pfrgc2bhrqqni-qtbase-5.12.0-bin", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/nix/store/7bvh4ydpza5cfhd7rw7pfrgc2bhrqqni-qtbase-5.12.0-bin/lib", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/nix/store/7bvh4ydpza5cfhd7rw7pfrgc2bhrqqni-qtbase-5.12.0-bin/lib/qt-5.12", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/nix/store/7bvh4ydpza5cfhd7rw7pfrgc2bhrqqni-qtbase-5.12.0-bin/lib/qt-5.12/plugins", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/nix/store/7bvh4ydpza5cfhd7rw7pfrgc2bhrqqni-qtbase-5.12.0-bin/lib/qt-5.12/plugins/platforms", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/nix/store/7bvh4ydpza5cfhd7rw7pfrgc2bhrqqni-qtbase-5.12.0-bin/lib/qt-5.12/plugins/platforms/libqminimal.so", {st_mode=S_IFREG|0555, st_size=149584, ...}) = 0

```

colemickens commented:
Well, now I can't get it to build. I switched it to cmake, per upstream, and added `qtkeychain` but I still get:

```
...
...
                                                                                                                                
--                           
-- =============================================================================                                                  
--                        libqtolm Build Information             
-- =============================================================================                                                  
-- Version: 0.1.0, API version: 0.1                              
-- Build type: Release                                                                                                          
-- Using compiler: GNU 7.4.0                                                                                                    
-- Using Qt 5.12.0 at /nix/store                                                                                                
-- Could NOT find Git (missing: GIT_EXECUTABLE)                  
-- Using system libolm                                           
-- =============================================================================                                                  
--                                                              
--                                                                                                                              
-- =============================================================================
--                        Quotient Build Information                                                                            
-- =============================================================================                     
-- Version: 0.6.0, API version: 0.6                              
-- Build type: Release                                                                                                          
-- Using compiler: GNU 7.4.0                                     
-- Install Prefix: /nix/store/mxnhxxp2d3ggrvxgbkzg2dqm29lff6vk-spectral-unstable-2019-08-15                                       
-- Using Qt 5.12.0 at /nix/store                                 
-- Could NOT find Git (missing: GIT_EXECUTABLE)                  
-- Using in-tree libQtOlm                                        
-- =============================================================================                                                  
--                                                               
CMake Error at CMakeLists.txt:89 (find_package):                 
  By not providing "FindQt5Keychain.cmake" in CMAKE_MODULE_PATH this project                                                      
  has asked CMake to find a package configuration file provided by                                                                
  "Qt5Keychain", but CMake did not find one.                     

  Could not find a package configuration file provided by "Qt5Keychain" with                                                      
  any of the following names:                                    

    Qt5KeychainConfig.cmake                                      
    qt5keychain-config.cmake                                     

  Add the installation prefix of "Qt5Keychain" to CMAKE_PREFIX_PATH or set                                                        
  "Qt5Keychain_DIR" to a directory containing one of the above files.  If                                                         
  "Qt5Keychain" provides a separate development package or SDK, be sure it                                                        
  has been installed.                                            


-- Configuring incomplete, errors occurred!                      
See also "/build/spectral-086891e/build/CMakeFiles/CMakeOutput.log".                                                              
See also "/build/spectral-086891e/build/CMakeFiles/CMakeError.log".                                                               
builder for '/nix/store/j6i3gibakdr6316z7382k0n8rwi8lbl4-spectral-unstable-2019-08-15.drv' failed with exit code 1 
```

lboklin commented:
Re: Qt5Keychain
`qtkeychain` defaults to qt4 (withQt5 ? false) for some reason, so overriding qtkeychain like this 
```
let qtkeychain-qt5 = qtkeychain.override { inherit qtbase qttools; withQt5 = true; };
in stdenv.mkDerivation ...
...
  buildInputs = [ qtkeychain-qt5 ...
```
fixes that issue.

Additionally you'll need to add `olm` to `buildInputs`.

colemickens commented:
@lboklin I'm looking at this now, but I'm a bit confused, this is what's in `pkgs/top-level/all-packages.nix`:

```
qtkeychain = callPackage ../development/libraries/qtkeychain {
  withQt5 = true;                                                               
};                  
```

So I don't see why the change you suggested would be necessary or have any effect?

lboklin commented:
> @lboklin I'm looking at this now, but I'm a bit confused, this is what's in `pkgs/top-level/all-packages.nix`:
> 
> ```
> qtkeychain = callPackage ../development/libraries/qtkeychain {
>   withQt5 = true;                                                               
> };                  
> ```
> 
> So I don't see why the change you suggested would be necessary or have any effect?

That's in `mkLibsForQt5`, not actually the top-level definition for `qtkeychain`, which doesn't set `withQt5 = true;`.

colemickens commented:
This is working for me now. Thanks @lboklin !

rnhmjoj commented:
This should be backported to 19.09, it's still broken.
